src/logger.cpp

```diff
@@ -0,0 +1,311 @@
+// Copyright 2023 The Forgotten Server Authors. All rights reserved.
+// Use of this source code is governed by the GPL-2.0 License that can be found in the LICENSE file.
+
+#include "otpch.h"
+
+#include "logger.h"
+
+#include <csignal>
+#include <iomanip>
+#include <spdlog/sinks/rotating_file_sink.h>
+#ifdef _WIN32
+#include <io.h>
+#define write _write
+#define STDERR_FILENO 2
+#else
+#include <unistd.h>
+#endif
+
+namespace {
+
+	std::mutex loggerMutex;
+	std::atomic<bool> loggerInitialized { false };
+	std::atomic<bool> shutdownInProgress { false };
+
+	spdlog::level::level_enum toSpd(LogLevel level) {
+		switch (level) {
+		case LogLevel::TRACE:
+			return spdlog::level::trace;
+		case LogLevel::DEBUG:
+			return spdlog::level::debug;
+		case LogLevel::INFO:
+			return spdlog::level::info;
+		case LogLevel::WARNING:
+			return spdlog::level::warn;
+		case LogLevel::ERRORR:
+			return spdlog::level::err;
+		case LogLevel::CRITICAL:
+			return spdlog::level::critical;
+		}
+		return spdlog::level::info;
+	}
+
+	LogLevel fromSpd(spdlog::level::level_enum level) {
+		switch (level) {
+		case spdlog::level::trace:
+			return LogLevel::TRACE;
+		case spdlog::level::debug:
+			return LogLevel::DEBUG;
+		case spdlog::level::info:
+			return LogLevel::INFO;
+		case spdlog::level::warn:
+			return LogLevel::WARNING;
+		case spdlog::level::err:
+			return LogLevel::ERRORR;
+		case spdlog::level::critical:
+			return LogLevel::CRITICAL;
+		default:
+			return LogLevel::INFO;
+		}
+	}
+
+	std::string generateLogFileName(std::string_view basePath) {
+		auto now = std::chrono::system_clock::now();
+		auto time_t = std::chrono::system_clock::to_time_t(now);
+
+		std::stringstream ss;
+		ss << std::put_time(std::localtime(&time_t), "%Y%m%d_%H%M%S");
+
+		std::filesystem::path path(basePath);
+		std::string directory = path.parent_path().string();
+		std::string baseName = path.stem().string();
+		std::string extension = path.extension().string();
+
+		if (!directory.empty()) {
+			try {
+				std::filesystem::create_directories(directory);
+			}
+			catch (const std::filesystem::filesystem_error& e) {
+				fprintf(stderr, "Failed to create log directory: %s\n", e.what());
+				throw;
+			}
+		}
+
+		return directory + "/" + baseName + "_" + ss.str() + extension;
+	}
+
+	bool checkDiskSpace(const std::string& path, size_t minSpaceBytes = 50 * 1024 * 1024) {
+		try {
+			auto space = std::filesystem::space(std::filesystem::path(path).parent_path());
+			return space.available > minSpaceBytes;
+		}
+		catch (const std::filesystem::filesystem_error&) {
+			return true;
+		}
+	}
+
+	class LogWithSpdLog final : public Logger {
+		public:
+			LogWithSpdLog(std::string_view filePath, size_t rotateSize, size_t rotateFiles) {
```

Comment:
This belongs to a free function `createLogger` that does all the preparation and then creates a new `LogWithSpdlog` when everything is checked